llama_index/schema.py

```diff
@@ -44,63 +85,125 @@ class BaseDocument(DataClassJsonMixin):
-            self.doc_hash = self._generate_doc_hash()
+        if self._id is None:
+            self._id = get_new_id(set())
+        if self.hash is None:
+            self.hash = self._generate_node_hash()
 
-        if self.extra_info is not None:
-            _validate_is_flat_dict(self.extra_info)
+        if self.metadata is not None:
+            _validate_is_flat_dict(self.metadata)
 
-    def _generate_doc_hash(self) -> str:
-        """Generate a hash to represent the document."""
-        doc_identity = str(self.text) + str(self.extra_info)
-        return sha256(doc_identity.encode("utf-8", "surrogatepass")).hexdigest()
+    @abstractmethod
+    def _generate_node_hash(self) -> str:
+        """Generate a hash to represent the node."""
 
     @classmethod
     @abstractmethod
     def get_type(cls) -> str:
-        """Get Document type."""
+        """Get Object type."""
 
-    @classmethod
-    def get_types(cls) -> List[str]:
-        """Get Document type."""
-        # TODO: remove this method
-        # a hack to preserve backwards compatibility for vector indices
-        return [cls.get_type()]
+    @abstractmethod
+    def get_content(self) -> str:
+        """Get object content."""
 
-    def get_text(self) -> str:
-        """Get text."""
-        if self.text is None:
-            raise ValueError("text field not set.")
-        return self.text
-
-    def get_doc_id(self) -> str:
-        """Get doc_id."""
-        if self.doc_id is None:
-            raise ValueError("doc_id not set.")
-        return self.doc_id
-
-    def get_doc_hash(self) -> str:
-        """Get doc_hash."""
-        if self.doc_hash is None:
-            raise ValueError("doc_hash is not set.")
-        return self.doc_hash
+    @abstractmethod
+    def metadata_str(self) -> str:
+        """Extra info string."""
+
+    @abstractproperty
+    @property
+    def is_content_none(self) -> bool:
+        """Check if text is None."""
 
     @property
-    def is_doc_id_none(self) -> bool:
+    def node_id(self) -> str:
+        if self._id is None:
+            raise ValueError("Node ID is not set!")
+        return self._id
+
+    @property
+    def is_id_none(self) -> bool:
         """Check if doc_id is None."""
-        return self.doc_id is None
+        return self.is_id_none is None
 
     @property
-    def is_text_none(self) -> bool:
-        """Check if text is None."""
-        return self.text is None
+    def source_node(self) -> Optional[RelatedNodeInfo]:
+        """Source object node.
+
+        Extracted from the relationships field.
+
+        """
+        if NodeRelationship.SOURCE not in self.relationships:
+            return None
+
+        relation = self.relationships[NodeRelationship.SOURCE]
+        if isinstance(relation, list): 
+            raise ValueError("Source object must be a single RelatedNodeInfo object")
+        return relation
+
+    @property
+    def prev_node(self) -> RelatedNodeInfo:
```

Comment:
I find it surprising that the `source_node` call returns None when not found, but the `prev_node` and `next_node` results in `ValueError`.

Reply:
Yea tbh I thought that was a little weird too while looking at the previous implementation. Probably safe to let it return `None` in all cases